scipy/stats/stats.py

```diff
@@ -4881,20 +5099,78 @@ def ks_2samp(data1, data2):
     data2 = np.sort(data2)
     n1 = data1.shape[0]
     n2 = data2.shape[0]
+    if min(n1, n2) == 0:
+        return Ks_2sampResult(1.0, 1.0)
+
     data_all = np.concatenate([data1, data2])
+    # using searchsorted solves equal data problem
+    # Strictly speaking, the cdfs are not correct at the repeated data
+    # values, but the difference between the two cdfs is correct.
     cdf1 = np.searchsorted(data1, data_all, side='right') / n1
     cdf2 = np.searchsorted(data2, data_all, side='right') / n2
-    d = np.max(np.absolute(cdf1 - cdf2))
-    # Note: d absolute not signed distance
-    en = np.sqrt(n1 * n2 / (n1 + n2))
-    try:
-        prob = distributions.kstwobign.sf((en + 0.12 + 0.11 / en) * d)
-    except Exception:
-        warnings.warn('This should not happen! Please open an issue at '
-                    'https://github.com/scipy/scipy/issues and provide the code '
-                    'you used to trigger this warning.\n')
-        prob = 1.0
+    cddiffs = cdf1 - cdf2
+    minS = -np.min(cddiffs)
+    maxS = np.max(cddiffs)
+    d = {'less': minS, 'greater': maxS, 'two-sided': max(minS, maxS)}[alternative]
+    g = gcd(n1, n2)
+    n1g = n1 // g
+    n2g = n2 // g
+    prob = -np.inf
+    if mode == 'auto':
+        if max(n1, n2) <= 100000:
+            mode = 'exact'
+        else:
+            mode = 'asymp'
+    elif mode == 'exact':
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        if n1g >= np.iinfo(np.int).max / n2g:
+            mode = 'asymp'
+
+    if mode == 'exact':
+        lcm = (n1 // g) * n2
+        h = int(np.round(d * lcm))
+        d = h * 1.0 / lcm
+        if h == 0:
+            prob = 1.0
+        else:
+            if alternative == 'two-sided':
+                if n1 == n2:
+                    prob = _compute_prob_outside_square(n1, h)
+                else:
+                    prob = 1 - _compute_prob_inside_method(n1, n2, g, h)
+                    print('prob=', prob)
+            else:
+                if n1 == n2:
+                    # binom(2n, n-h) / binom(2n, n)
+                    # Evaluating in that form incurs roundoff errors from special.binom
+                    # Instead calculate directly
+                    prob = 1.0
+                    for j in range(h):
+                        prob = (n1 - j) * prob / (n1 + j + 1)
+                else:
+                    try:
+                        num_paths = _count_paths_outside_method(n1, n2, g, h)
+                        bin = special.binom(n1 + n2, n1)
+                        if not np.isfinite(bin) or not np.isfinite(num_paths) or num_paths > bin:
+                            raise FloatingPointError()
+                        prob = num_paths / bin
+                    except FloatingPointError:
+                        pass
+    if not np.isfinite(prob) or prob > 1 or prob < 0:
```

Comment:
should one warn the user that s.th. went wrong?

Reply:
This branch is entered if either a) mode !='exact', or b) mode=='exact' and the computation in `_count_paths_outside_method()` overflowed; or c) mode=='exact' and or `_count_paths_outside_method` incurred roundoff error putting the result outside of the interval `[0,1]`.
(`_compute_prob_inside_method()` and `_compute_prob_outside_square()` can be done without overflowing by always working with probs, but `_count_paths_outside_method` has to compute the number of legal paths, and this can involve very large binomial coefficients.)
In case a) there is no error, as no computation has been done.
If the sample sizes are large enough to cause case b) or c), then switching to `asymp` should give a decent approximation to the true probability.  I've reworked to make clearer and raise a warning in cases b) and c) if the passed in value of mode is 'exact'.

Reply:
makes sense, thanks for the clarification